Gemfile

```diff
@@ -1,16 +1,17 @@
 source "https://rubygems.org"
 git_source(:github){|repo| "https://github.com/#{repo}.git"}
 ruby "2.7.1"
+gem "bcrypt", "3.1.13"
```

Comment:
Gem này có tác dụng gì vậy em?

Reply:
Gem bcrypt là gem sử dụng thuật toán mã hóa Blowfish để mã hóa mật khẩu thành chuỗi password_digest, đây là mã hóa một chiều và không thể lấy lại khi đã biết chuỗi mật khẩu trong database. Bcrypt hỗ trợ đầu vào của mỗi ký tự ở định dạng UTF-8, vì vậy nó có thể băm mật khẩu mà không giới hạn ký tự cho các ký tự cả chữ và số ạ.

Reply:
Cách nó kiểm tra mật khẩu như thế nào em

Reply:
Khi user thiết lập password, bcrypt tạo ra một chuỗi salt ngẫu nhiên (random theo từng user), kết hợp với password của user để tạo ra chuỗi hash cuối cùng lưu vào trường encrypted_password dưới db.
Khi user đăng nhập tài khoản, hệ thống sẽ tìm user name hoặc email tương ứng trong db để lấy encrypted_password. Sau đó lấy ra chuỗi salt (29 ký tự đầu tiên trong encrypted_password).
bcrypt sẽ sử dụng chuỗi salt đó với password mà người dùng đăng nhập để tạo ra một chuỗi hash và so sánh chuỗi hash vừa tạo đó với encrypted_password. Nếu đúng thì xác thực thành công.